File: Java/Migration/unit-tests-csharp/unit-tests-csharp/P04NSubstitute/Ex02MultipleReturnsAndOrderVerification.cs
using NSubstitute;
using NUnit.Framework;
using TddEbook.TddToolkit;
using unit_tests_csharp.P04NSubstitute.Ex010203ProductionCode;

namespace unit_tests_csharp.P04NSubstitute
{
  public class Ex02MultipleReturnsAndOrderVerification
  {
    [Test]
    public void ShouldReturnMultipleDataElements()
    {
      //GIVEN
      var source = Substitute.For<IDataSource>();
      var data1 = Any.Instance<Data>();
      var data2 = Any.Instance<Data>();

      source.RetrieveData().Returns(data1, data2);

      //WHEN
      var result1 = source.RetrieveData();
      var result2 = source.RetrieveData();

      //THEN
      Assert.AreEqual(data1, result1);
      Assert.AreEqual(data2, result2);
    }

    [Test] //order verification
    public void ShouldCopyDataFromSourceToDestinationMultipleTimes()
    {
      //GIVEN
      var copyOperation = new CopyOperation();
      var destination = Substitute.For<IDataDestination>();
      var source = Substitute.For<IDataSource>();
      var data1 = Any.Instance<Data>();
      var data2 = Any.Instance<Data>();


      source.RetrieveData().Returns(data1, data2);

      //WHEN
      copyOperation.ApplyTo(source, destination);
      copyOperation.ApplyTo(source, destination);

      //THEN
      Received.InOrder(() =>
      {
        destination.Save(data1);
        destination.Save(data2);
      });
    }
  }
}

File: Java/xunit-test-patterns/src/test/java/_09_mock_objects/Ex02MultipleReturnsAndOrderVerification.java
package _09_mock_objects;

import autofixture.publicinterface.Any;
import lombok.val;
import org.mockito.InOrder;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.inOrder;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class Ex02MultipleReturnsAndOrderVerification {
    @Test
    public void shouldReturnMultipleDataElements() {
        //GIVEN
        val source = mock(DataSource.class);
        val data1 = Any.anonymous(Data.class);
        val data2 = Any.anonymous(Data.class);

        when(source.retrieveData())
            .thenReturn(data1)
            .thenReturn(data2);

        //WHEN
        val result1 = source.retrieveData();
        val result2 = source.retrieveData();

        //THEN
        assertThat(result1).isEqualTo(data1);
        assertThat(result2).isEqualTo(data2);
    }

    @Test //order verification
    public void shouldCopyDataFromSourceToDestinationMultipleTimes() {
        //GIVEN
        CopyOperation copyOperation = new CopyOperation();
        DataDestination destination = mock(DataDestination.class);
        DataSource source = mock(DataSource.class);
        Data data1 = Any.anonymous(Data.class);
        Data data2 = Any.anonymous(Data.class);


        when(source.retrieveData()).thenReturn(data1).thenReturn(data2);

        //WHEN
        copyOperation.applyTo(source, destination);
        copyOperation.applyTo(source, destination);

        //THEN
        InOrder inOrder = inOrder(destination);
        inOrder.verify(destination).save(data1);
        inOrder.verify(destination).save(data2);
    }
}
